### ISSUE TITLE
Adds a New Patch Policy "Remove" to synchronize with the status of the managed resource

### ISSUE THREAD

**Issue by nueavv** [crossplane/crossplane#3542]
### What problem are you facing?

I made XR and Composition with provider_kubernetes to create a deployment. and update the status of a deployment to the XR. 
Since the status of deployment has has two states "status.unavailableReplicas", "status.availableReplicas", the status of XR is able to have two field. like this. 

```
  nginx:
    availableReplicas: 1
    unavailableReplicas: 1
    conditions:
      - lastTransitionTime: '2022-12-21T04:45:08Z'
        lastUpdateTime: '2022-12-21T04:45:08Z'
        message: Deployment has minimum availability.
        reason: MinimumReplicasAvailable
        status: 'True'
        type: Available
      - lastTransitionTime: '2022-12-21T04:44:54Z'
        lastUpdateTime: '2022-12-21T04:45:08Z'
        message: ReplicaSet "abcd-55fb695654" has successfully progressed.
        reason: NewReplicaSetAvailable
        status: 'True'
        type: Progressing
    replicas: 1
    updatedReplicas: 1
```

### How could Crossplane help solve your problem?

Add Patch Policy "Remove" to remove a tofieldpath when the value of fromfieldpath is nill.

```
        - type: ToCompositeFieldPath
          fromFieldPath: status.atProvider.manifest.status
          toFieldPath: status.nginx
          policy:
            toFieldPath: Remove # <-- New Policy 
```

**Comment by nueavv** [comment]
#3543

**Comment by bobh66** [comment]
Can you provide some more details about what the actual problem is?  You explained the structure of the Kubernetes deployment, but it's not clear what the problem is and how your proposed changes will resolve it.  It would help to have an example of what the "bad" data looks like and what you expect the "good" data to look like, and tie those together with how your proposed changes work.   Thanks!

**Comment by bobh66** [comment]
Thanks @nueavv - that makes sense.  I think it would be better for the Deployment controller to always populate both status fields with the correct value - instead of not populating or removing a field when it's value it 0, the field should exist with the value explicitly set to 0.  That would be much clearer to the consumer and much easier to consume.

Did you consider requesting that provider-kubernetes always copy the exact status from the target kubernetes Object?  That should cause the fields to be removed when they disappear, and as a user I would expect that the provider-kubernetes status.atProvider.manifest.status should always exactly match the status of the remote object.   If the extra attributes are being left in status.atProvider.manifest.status instead of being removed that seems like a bug in provider-kubernetes.

I think fixing that would make for a better user experience than needing to explicitly remove a field when the object is used.

**Comment by nueavv** [comment]
I add a new transform policy "remove" because the object created by provider_kubernetes synchronized well with the deployment status. provider_kubernetes deletes the unavailableReplicas if the deployment wouldn't have it.

And  XR and XRC are supposed to be the same, followed by composition. 
But It isn't if I create a composition like this.
```
        - type: ToCompositeFieldPath
          fromFieldPath: status.atProvider.manifest.status
          toFieldPath: status.nginx
```

If the deployment is set to the replicas 0, the status of it would be like the below. as you can see, there are no unavailableReplicas, and availableReplicas. either. 
```
status:
  conditions:
  - lastTransitionTime: "2022-12-23T05:16:27Z"
    lastUpdateTime: "2022-12-23T05:16:27Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2022-12-23T05:16:15Z"
    lastUpdateTime: "2022-12-23T05:16:27Z"
    message: ReplicaSet "cloudtechweb-55fb695654" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 2
```

I think deleting empty fields is needed, even if we don't add the new policy "remove".

**Comment by ytsarev** [comment]
@nueavv just to avoid confusion, in the above example and in #3546 we are talking about PatchPolicy, not a transform, is that correct?

**Comment by nueavv** [comment]
Hi, @ytsarev  yes, you're right. I should edit that example.

**Comment by bobh66** [comment]
@nueavv - thanks for the explanation, I'm still trying to wrap my head around the problem scenario.

Using your patch example from above:

```
        - type: ToCompositeFieldPath
          fromFieldPath: status.atProvider.manifest.status
          toFieldPath: status.nginx
```

I assume that the first time the patch is executed the status matches between the two objects and looks something like this:

```
  nginx:
    unavailableReplicas: 1
    conditions:
      - lastTransitionTime: '2022-12-21T04:45:08Z'
        lastUpdateTime: '2022-12-21T04:45:08Z'
        message: Deployment has minimum availability.
        reason: MinimumReplicasAvailable
        status: 'False'
        type: Available
      - lastTransitionTime: '2022-12-21T04:44:54Z'
        lastUpdateTime: '2022-12-21T04:45:08Z'
        message: ReplicaSet "abcd-55fb695654" has successfully progressed.
        reason: NewReplicaSetAvailable
        status: 'True'
        type: Progressing
    replicas: 1
```

Then on a subsequent reconciliation when the patch is executed again it will copy the updated version of _status.atProvider.manifest.status_ to _status.nginx_, which we might expect to look something like this (ignore the timestamps):

```
  nginx:
    availableReplicas: 1
    conditions:
      - lastTransitionTime: '2022-12-21T04:45:08Z'
        lastUpdateTime: '2022-12-21T04:45:08Z'
        message: Deployment has minimum availability.
        reason: MinimumReplicasAvailable
        status: 'True'
        type: Available
      - lastTransitionTime: '2022-12-21T04:44:54Z'
        lastUpdateTime: '2022-12-21T04:45:08Z'
        message: ReplicaSet "abcd-55fb695654" has successfully progressed.
        reason: NewReplicaSetAvailable
        status: 'True'
        type: Progressing
    replicas: 1
```
where the unavailableReplicas has been removed (instead of setting it to 0 which would make more sense to me) and has been replaced by the availableReplicas field set to 1.

Based on the example you provided above it appears that the patch operation that copies the status from the provider-kubernetes Object resource is overlaying the new status object on top of the old one (or merging the objects) instead of replacing the object entirely.

If this is indeed the problem, I would have expected the existing mergeOptions.keepMapValues default value of "false" to have prevented this from happening, based on the comment here: https://github.com/crossplane/crossplane-runtime/blob/master/apis/common/v1/merge.go#L33

It seems like the default should be to replace the status.nginx object with the current status.atProvider.manifest.status object, which would eliminate the need for the new patch policy.   @negz - am I correct about the way the patch should work by default?

**Comment by nueavv** [comment]
@bobh66 , I agree with your opinion, But keepMapValues policy doesn't work for this. In this code, it says that it overwrites only the non-empty fields. 
```
// https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/claim/configurator.go#L191

	if err := merge(ucm.Object["status"], ucp.Object["status"],
		// Status fields from composite overwrite non-empty fields in claim
		withMergeOptions(mergo.WithOverride),
		withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceStatusProps())...)); err != nil {
		return errors.Wrap(err, errMergeClaimStatus)
	}
```

And to merge the composite status, Mergo is used. It doesn't have a merge function removing empty fields. 
https://pkg.go.dev/github.com/imdario/mergo#pkg-functions

**Comment by bobh66** [comment]
Thanks @nueavv - that makes sense.  It looks like the "omitempty" attribute is specified for these fields in the apps DeploymentStatus struct, which causes them to be removed during JSON marshalling when the values are 0.

I wonder if changing the withMergoOptions() call to use merge.WithOverwriteWithEmptyValue - https://pkg.go.dev/github.com/imdario/mergo#WithOverwriteWithEmptyValue - would be a better solution?  If we really want the Claim status to match the Composite status (and I think we do) then it should overwrite with empty values as needed (and possibly WithOverrideEmptySlice too?  https://pkg.go.dev/github.com/imdario/mergo#WithOverrideEmptySlice)

I'm not trying to be difficult, I just think that the root cause of this scenario is an artifact of how the JSON data structures are defined and therefore is an internal implementation detail that the user should not have to think about, and definitely should not have to adjust their patching policies to account for.

@negz  - thoughts?

**Comment by nueavv** [comment]
@bobh66, Thanks for your help.
I also think it would be better to match the status of the managed resource without any policy. 

When XR is patching with an optional field deleted, It just skips.
I thought there would be some reason for doing nothing with it.

```
// https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/patches.go#L170
	in, err := fieldpath.Pave(fromMap).GetValue(*p.FromFieldPath)
	if IsOptionalFieldPathNotFound(err, p.Policy) {
		return nil
	}
	if err != nil {
		return err
	}
```

And I tried WithOverwriteWithEmptyValue, But it doesn't remove the empty field.
this is the code I tested.
```
	if err := merge(ucm.Object["status"], ucp.Object["status"],
		// Status fields from composite overwrite non-empty fields in claim
		withMergeOptions(mergo.WithOverwriteWithEmptyValue),
		withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceStatusProps())...)); err != nil {
		return errors.Wrap(err, errMergeClaimStatus)
	}
```

The differences between WithOverride and WithOverwriteWithEmptyValue are below.
```
// WithOverride will make merge override non-empty dst attributes with non-empty src attributes values.
func WithOverride(config *Config) {
	config.Overwrite = true
}

// WithOverwriteWithEmptyValue will make merge override non empty dst attributes with empty src attributes values.
func WithOverwriteWithEmptyValue(config *Config) {
	config.Overwrite = true
	config.overwriteWithEmptyValue = true
}

```

**Comment by negz** [comment]
If I understand the problem correctly, I think I agree with @bobh66 here. At least, I would definitely expect claim status to effectively be a copy of XR status (as opposed to have XR status overlaid on claim status).

It looks like merging the XR status to the claim status was added in https://github.com/crossplane/crossplane/pull/2077. In retrospect I'm not really sure why. We expect claims and XRs to effectively mirror each other, so just copying XR status -> claim status (while preserving any claim-specific machinery fields) should be fine.

Something that's not clear to me here is whether there's also a MR -> XR patching issue that we're facing here, though. I would hope that if you patch an entire object (e.g. MR `status.atProvider.manifest.status` -> XR `status.nginx`) the _default_ behaviour would be to wholesale replace rather than try to merge those objects, presuming there was no merge config supplied.

CC @turkenh who I believe built the merge patch policy.

**Comment by nueavv** [comment]
Hi, @negz and @bobh66. I agree with your opinion.
So I create New PR, which copies the XR status from the MR status and the XRC status from the XR status.

**Comment by nueavv** [comment]
Hi, @negz. These are the problems I faced.
1. there is a removed status field on MR.
2. when the non-exists field is found on XR., they do nothing.
```
	if IsOptionalFieldPathNotFound(err, p.Policy) {
		return nil
	}
```
4. Also, Merging the XRC status doesn't remove that field too.
```
	if err := merge(ucm.Object["status"], ucp.Object["status"],
		// Status fields from composite overwrite non-empty fields in claim
		withMergeOptions(mergo.WithOverride),
		withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceStatusProps())...)); err != nil {
		return errors.Wrap(err, errMergeClaimStatus)
	}
```

**Comment by nueavv** [comment]
@negz. It is weird, though. Since the default behavior of the FromCompositeFieldPath is copying. The default behavior of the  ToCompositeFieldPath should also be the same, not removing, even if they are not found on MR.
But I agree that the two statuses of those XR/XRC should be the same without any policy.

**Comment by nueavv** [comment]
Hello, @negz  @bobh66 what should i do ?

**Comment by jbw976** [comment]
Looks like this issue is now being addressed by PR #3596 and further input is being requested of the PR author there 😉

**Comment by pedjak** [comment]
Fixed via https://github.com/crossplane/crossplane/pull/4896.
